src/gemdat/transitions.py

```diff
@@ -448,34 +458,49 @@ def _calculate_atom_states(
         The value corresponds to the index in the `site_coords`.
         -1 indicates that atom is not at any site.
     """
-    # Unit cell parameters
+
+    def _site_radius_iterator():
+        for label, radius in site_radius.items():
+            if label:
+                grouped = ((k, site) for k, site in enumerate(sites)
+                           if site.label == label)
+                key, site_group = zip(*grouped)
+                frac_coords = np.array(
+                    [site.frac_coords for site in site_group])
+                yield frac_coords, np.array(key), radius
+            else:
+                yield sites.frac_coords, None, radius
+
     lattice = trajectory.get_lattice()
 
-    site_coords = sites.frac_coords
+    cutoff = max(list(site_radius.values()))
+
+    traj_frac_coords = trajectory.positions.reshape(-1, 3)
+    traj_cart_coords = np.dot(traj_frac_coords, lattice.matrix)
 
-    # Input array with site coordinates [site, (x, y, z)]
-    site_cart_coords = np.dot(site_coords, lattice.matrix)
     site_coords_tree: PeriodicKDTree = PeriodicKDTree(
         box=np.array(lattice.parameters, dtype=np.float32))
```

Comment:
Have you tried setting the leafsize keyword (default 10) to see if a better performance is possible?

Reply:
Good point! Seems to make very little difference, so let's leave it at the default. 👍